### PR TITLE
Add keycloak_authz_permission module

### PR DESCRIPTION
**Pull request by mattock**
##### SUMMARY

This new module can be used to create, update and remove Keycloak client authorization permissions. The client needs to have authorization enabled for this module to make sense. Both resource- and scope-based authorization permission are supported.

There are some peculiarities with the JSON payloads and endpoints:
* GET requests must be pointed to the "policy" endpoint. The returned data is in "policy" format and lacks information about attached resources, scopes and policies and includes an empty "config" parameter.
* POST and PUT requests must be pointed to the "permission" endpoints. The JSON payload must contain resource, scope and policy information.

The  end result is that it is not possible to query the current state of resources, scopes or policies attached to a permission. This module therefore just applies the current state (JSON payload) on existing permissions on every run. This is likely less harmful than not doing anything and letting the configuration diverge due to changes done outside of Ansible.  

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

keycloak_authz_permission

##### ADDITIONAL INFORMATION

The first commits in this PR add keycloak_authz_authorization_scope module and other plumbing required. The main reason for them to exist is to allow the integration tests to work. Those commits must be removed once https://github.com/ansible-collections/community.general/pull/6256 is merged. Any issues in that commit/module can be safely ignored.

### CODE REVIEW

File: plugins/modules/keycloak_authz_authorization_scope.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_authorization_scope.py:68]:
```suggestion
    - community.general.keycloak
    - community.general.attributes
```
You also need to declare whether the module supports `check_mode` and `diff_mode` (see all other modules for examples how that works).

File: plugins/modules/keycloak_authz_authorization_scope.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_authorization_scope.py:95]:
```suggestion
    description: Representation of the authorization scope after module execution.
```

File: plugins/modules/keycloak_authz_authorization_scope.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_authorization_scope.py:100]:
```suggestion
            description: ID of the authorization scope.
```
See below for more similar cases.

File: plugins/modules/keycloak_authz_authorization_scope.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_authorization_scope.py:30]:
```suggestion

version_added: 6.6.0

```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments.

File: plugins/modules/keycloak_authz_authorization_scope.py
Comment (mattock) [reply plugins/modules/keycloak_authz_authorization_scope.py:30]: Please ignore anything code related to authorization scopes in this PR. That code is only included because the integration tests require it. I will rebase this PR on top of https://github.com/ansible-collections/community.general/pull/6256 once that PR is merged.

**Comment by mattock** [comment]
The latest commit fixes the same type of stylistic issues in *keycloak_authz_permission.py* that @felixfontein noticed in https://github.com/ansible-collections/community.general/pull/6256.

**Comment by mattock** [comment]
The diff/check mode issue should now be fixed.

**Comment by felixfontein** [comment]
Can you please rebase now that the one module has been merged, and apply adjustments to the new module similar to the changes you had to made to the other? Thanks :)

**Comment by mattock** [comment]
@felixfontein I rebased this PR on top of HEAD of "main". Locally all validation and unit tests pass, but let's see about GitHub Actions. I also made all the same documentation fixes here as in the keycloak_authz_authorization_scope PR earlier. 

I can't implement full check_mode in this module, because I'm unable to query the current state. I could support a partial check_mode that just reports back on permission presence and absence. Do you think doing that would be worth the effort?

**Comment by mattock** [comment]
@felixfontein all the validation and tests now pass. Awaiting for feedback on check_mode and other things :smile:.

**Comment by mattock** [comment]
Ping @felixfontein :smile:.

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:14]:
```suggestion
module: keycloak_authz_permission

version_added: 7.0.0
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:444]: If anything is updated, `changed` must be set to `True`. So either the above function call isn't needed, or the assignment to `False` is wrong.

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:112]: I'm not sure that this option should exist. Such functionality should be provided by an `_info` module, not by the module itself.

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:43]: All other keycloak modules support check mode, and almost all others also support diff mode. Is there a technical reason why not at least check mode can be implemented here?

**Review by felixfontein — COMMENTED** [review]
Sorry for the delay!

File: plugins/modules/keycloak_authz_permission.py
Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:446]: Let me explain this - this is also tangentially related to check_mode. Keycloak authorization permission are handled "interestingly". When you create or remove a permission, you use a "permission endpoint" or URL. When you query existing permissions, you actually query the "policies endpoint". To make matters worse, these two endpoints return or expect different type of data:

* Data from the *policies* endpoint does not include information about resources, policies or scopes (all lists)
* Data from *policies* endpoint is returned in a different format than what the *permission* endpoints expect when creating authorization permissions

The end result is that it is not possible to detect the current state of the permission. This means there are two options:

1. Only support "create" and "delete". Ignore any attempts to manage the desired state of created objects.
1. Always update all objects that already exists. This maintains the desired state but it is impossible to know if the update was actually needed or not.
 
I would prefer option 2) as there's no real harm in doing it afaics. Whether *result['changed']* should be False or True or None or what I don't know. That more about Ansible conventions I guess than anything else.

Option 1) would be, in my opinion, more surprising and limit the usability of this module.

Thoughts?

File: plugins/modules/keycloak_authz_permission.py
Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:114]: Do you have any pointers / examples on how to create an info module for this kind of use-case? This is the first time I heard about it :smile:.

File: plugins/modules/keycloak_authz_permission.py
Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:45]: I can implement check_mode for create and remove operations. What I cannot do is check if changes are required to the desired state of already existing objects. The full explanation is in a comment above.

I guess the discussion revolves around what is the correct value for *result['changed']* when you cannot know for sure if something changed (or would change).

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [reply plugins/modules/keycloak_authz_permission.py:446]: Option 2) is fine for me, but this needs to be documented, and `changed=True` should always be returned when you aren't sure whether something was changed.

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [reply plugins/modules/keycloak_authz_permission.py:114]: See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#creating-an-info-or-a-facts-module for example.

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [reply plugins/modules/keycloak_authz_permission.py:45]: Check mode with the same behavior as the real implementation (i.e. always return `changed=True` if you're not sure) is already good enough IMO.

Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:45]: Thanks for the feedback @felixfontein. I'll make the necessary changes.

Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:45]: Fixed the `result['changed']` issue and added support for check_mode. Info module next.

Comment (mattock) [reply plugins/modules/keycloak_authz_permission.py:45]: I implemented the "query_state" as an info module and ripped away that functionality from the main module. I also made the integration tests use the info module.

File: tests/integration/targets/keycloak_authz_permission/aliases
Comment (felixfontein) [inline tests/integration/targets/keycloak_authz_permission/aliases:5]:
```suggestion
unsupported
keycloak_authz_permission_info
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:16]:
```suggestion
version_added: 7.1.0
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:18]:
```suggestion
short_description: Allows administration of Keycloak client authorization permissions via Keycloak API
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:39]:
```suggestion
      U(https://www.puppeteers.net/blog/keycloak-authorization-services-rest-api-paths-and-payload/)
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:98]:
```suggestion
            - Resource-based permissions cannot have scopes attached to them.
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:16]:
```suggestion
version_added: 7.1.0
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:18]:
```suggestion
short_description: Query Keycloak client authorization permissions information
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:32]:
```suggestion
      U(https://www.puppeteers.net/blog/keycloak-authorization-services-rest-api-paths-and-payload/)
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:39]:
```suggestion
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:60]:
```suggestion
    - community.general.attributes
    - community.general.attributes.info_module
```

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:68]:
```suggestion
  community.general.keycloak_authz_permission_info:
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:132]:
```suggestion
  community.general.keycloak_authz_permission:
```

File: plugins/modules/keycloak_authz_permission.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission.py:149]:
```suggestion
  community.general.keycloak_authz_permission:
```

**Review by felixfontein — COMMENTED** [review]
I've added some more comments on the docs. I think that besides these, this is OK for merging.

**Comment by mattock** [comment]
All changes looked good to me. As the PR will be squashed I let GitHub make the commits on my behalf.

**Comment by felixfontein** [comment]
Please note that you missed three comments - GitHub unfortunately hides review comments if there are more than 10 in one review. You need to explicitly press on the 'load more' link under '3 hidden conversations' in the UI... (And yes, I think this is stupid. I sometimes even do two reviews in a row to avoid having more than 10, because in half of the times folks don't notice the hidden ones... I wish GH would fix this...)

**Comment by mattock** [comment]
> Please note that you missed three comments - GitHub unfortunately hides review comments if there are more than 10 in one review. You need to explicitly press on the 'load more' link under '3 hidden conversations' in the UI... (And yes, I think this is stupid. I sometimes even do two reviews in a row to avoid having more than 10, because in half of the times folks don't notice the hidden ones... I wish GH would fix this...)

Oh, I did not know about that :smile: . Anyhow, all your suggestions have been committed now.

**Review by felixfontein — APPROVED** [review]
I think this looks good (though I don't know Keycloak), so I'll merge this in a few days if nobody objects.

(Note that I will first merge #6476, at which point there will likely be conflicts in the keycloak module utils. These should be easy to fix though, both PRs append code at the bottom.)

**Comment by felixfontein** [comment]
@mattock unfortunately there are now conflicts, can you resolve them (with a rebase)? This should be pretty easy since it's mainly the order of stuff added to the keycloak module utils that is likely a problem.

**Comment by felixfontein** [comment]
@mattock I cannot push to your branch so I cannot do the rebase for you. (If you don't have time and don't mind me rebasing, feel free to tick that checkbox...)

**Comment by mattock** [comment]
@felixfontein took a while, but I've done the (trivial) rebase now.

**Comment by felixfontein** [comment]
Thanks!

Could it be that several changes got undone? I see a lot of resolved comments of mine where the things I commented on haven't been changed (anymore? I think they were changed at some point).

Also, `version_added` needs to be bumped to 7.2.0.

**Comment by felixfontein** [comment]
Actually if you look at the changes due to the rebase: https://github.com/ansible-collections/community.general/compare/6d6e430b0c88bb0ff741e8a8aec048fa2c83d06c..5cc78b96109d9ce8a2077b59f22c70db32e710d1 you can see that some changes apply also for the newly added files - these changes basically undo the things mentioned above.

**Comment by mattock** [comment]
@felixfontein I've never noticed a simple rebase destroy data. That said, I did not back up my earlier branch, so I can't be sure. Anyhow, I believe all the fixes are now in (again).

File: plugins/modules/keycloak_authz_permission_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_authz_permission_info.py:38]: This one was gone as well:
```suggestion
```

**Review by felixfontein — COMMENTED** [review]
Besides this one (and obviously the updated `version_added` fields), everything is as before.

**Comment by felixfontein** [comment]
@mattock there is a new relesae (7.2.0) on Monday, please remove the indicated part and update `version_added` before Monday.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@mattock thanks for your contribution!

**Pull request merged**
